pkg/security/ptracer/proc.go

```diff
@@ -0,0 +1,286 @@
+				return true
+			}
+
+			if proc = processes[ppid]; proc == nil {
+				return true
+			}
+		}
+	}
+
+	for pid, proc := range processes {
+		if isTraced(proc) {
+			traced = append(traced, pid)
+		}
+	}
+
+	// remove traced from the process list
+	for _, pid := range traced {
+		delete(processes, pid)
+	}
+
+	return processes, nil
+}
+
+func pidTTY(pid int32) string {
+	fdPath := fmt.Sprintf("/proc/%d/fd/0", pid)
+
+	ttyPath, err := os.Readlink(fdPath)
+	if err != nil {
+		return ""
+	}
+
+	if ttyPath == "/dev/null" {
+		return ""
+	}
+
+	if strings.HasPrefix(ttyPath, "/dev/pts") {
+		return "pts" + path.Base(ttyPath)
+	}
+
+	if strings.HasPrefix(ttyPath, "/dev") {
+		return path.Base(ttyPath)
+	}
+
+	return ""
+}
+
+// MaxEnvVarsCollected max env to be collected
+const MaxEnvVarsCollected = 256
+
+func zeroSplitter(data []byte, atEOF bool) (advance int, token []byte, err error) {
+	for i := 0; i < len(data); i++ {
+		if data[i] == '\x00' {
+			return i + 1, data[:i], nil
+		}
+	}
+	if !atEOF {
+		return 0, nil, nil
+	}
+	return 0, data, bufio.ErrFinalToken
+}
+
+func newEnvScanner(f *os.File) (*bufio.Scanner, error) {
+	_, err := f.Seek(0, io.SeekStart)
+	if err != nil {
+		return nil, err
+	}
+
+	scanner := bufio.NewScanner(f)
+	scanner.Split(zeroSplitter)
+
+	return scanner, nil
+}
+
+func envVars(pid int32) []string {
+	filename := fmt.Sprintf("/proc/%d/environ", pid)
+
+	f, err := os.Open(filename)
+	if err != nil {
+		return nil
+	}
+	defer f.Close()
+
+	// first pass collecting only priority variables
+	scanner, err := newEnvScanner(f)
+	if err != nil {
+		return nil
+	}
+
+	var envs []string
+	envCounter := 0
+
+	for scanner.Scan() {
+		text := scanner.Text()
+		if len(text) > 0 {
+			envs = append(envs, text)
+
+			envCounter++
+			if envCounter > MaxEnvVarsCollected {
+				break
+			}
+		}
+	}
+
+	return envs
+}
+
+func procToMsg(proc *process.Process) (*ebpfless.Message, error) {
+	createTime, err := proc.CreateTime()
+	if err != nil {
+		return nil, err
+	}
+
+	uids, err := proc.Uids()
+	if err != nil {
+		return nil, err
+	}
+
+	gids, err := proc.Gids()
+	if err != nil {
+		return nil, err
+	}
+
+	cmdLine, err := proc.CmdlineSlice()
+	if err != nil {
+		return nil, err
+	}
+
+	filename, err := os.Readlink(fmt.Sprintf("/proc/%d/exe", proc.Pid))
+	if err != nil {
+		return nil, fmt.Errorf("snapshot failed for %d: couldn't readlink binary: %w", proc.Pid, err)
+	}
+	if filename == "/ (deleted)" {
+		return nil, fmt.Errorf("snapshot failed for %d: binary was deleted", proc.Pid)
+	}
+
+	return &ebpfless.Message{
+		Type: ebpfless.MessageTypeSyscall,
+		Syscall: &ebpfless.SyscallMsg{
+			Type:      ebpfless.SyscallTypeExec,
+			PID:       uint32(proc.Pid),
+			Timestamp: uint64(time.Unix(0, createTime*int64(time.Millisecond)).UnixNano()),
+			Exec: &ebpfless.ExecSyscallMsg{
+				File: ebpfless.FileSyscallMsg{
+					Filename: filename,
+				},
+				Args: cmdLine,
+				Envs: envVars(proc.Pid),
+				TTY:  pidTTY(proc.Pid),
```

Comment:
We could use `getPidTTY()` instead of a new func doing the ~same thing.

Reply:
I'll do